scanner.coffee

```diff
@@ -86,25 +105,49 @@ scan = (line, linenumber, tokens) ->
                 pos += 2
 
             else if /\"/.test line[pos]
-                pos++ until /[^\\]\"/.test(line.substring pos, pos + 2)
-                emit 'STRLIT', line.substring ++start, pos + 1
+                pos++ until /[^\\]\"|\$\(/.test(line.substring pos, pos + 2)
+                if /\$\(/.test(line.substring pos, pos + 2)
+                    emit 'STRPRT', line.substring ++start, pos
+                    start = pos
+                    emit '$('
+                    interpolating = true
+                else
+                    emit 'STRLIT', line.substring ++start, pos + 1
                 pos += 2
 
             # One-character tokens
-            else if /[+\-*\/(),:=<>\{\}\^&\|!]|(?:\.[^0-9])/.test(line.substring(pos, pos + 2))
+            else if /^(?:[+\-*\/(),:=<>\[\]\{\}\^&\|!]|(?:\.[^0-9]))/.test(line.substring(pos, pos + 2))
                 emit line[pos++]
+                if interpolating
+                    interpolatingDepth++ if line[pos - 1] is '('
+                    if line[pos - 1] is ')'
+                        if interpolatingDepth isnt 0
+                            interpolatingDepth--
+                        else
+                            start = pos
+                            interpolating = false
+                            pos++ until /[^\\]\"|\$\(/.test(line.substring pos, pos + 2)
+                            if /\$\(/.test(line.substring pos, pos + 2)
+                                emit 'STRPRT', line.substring start, pos + 1
+                                emit '$('
+                                interpolating = true
+                            else
+                                emit 'STRPRT', line.substring start, pos + 1
+                            pos += 2
```

Comment:
@rtoal: Advice on better ways to do this? I don't like the duplicated code here, but I need access to the state of the scanner.

I'm thinking of breaking it out into a function like `emit`. Thoughts?